rfcs/0000-user-defined-aggregate-functions.md

```diff
@@ -0,0 +1,269 @@
+
+```mermaid
+sequenceDiagram
+    participant UP as Upstream<br/>Downstream
+    participant AG as Agg Executor
+    participant CP as UDF Server
+    UP ->> AG: init(epoch)
+    Note over AG: load state<br/>from state table
+    AG ->>+ CP: doExchange(fid, state)
+    CP -->> AG: output @ epoch
+    Note right of CP: decode state,<br/>create accumulator
+    loop each epoch
+        loop each chunk
+            UP ->> AG: input chunk
+            AG ->> CP: input chunk
+            Note right of CP: accumulate,<br/>retract
+        end
+        UP ->> AG: barrier(epoch+1)
+        AG ->> CP: finish
+        Note right of CP: encode state,<br/>get output
+        CP -->> AG: state + output @ epoch+1
+        Note over AG: store state<br/>into state table
+        AG ->> UP: - output @ epoch<br/>+ output @ epoch+1
+        AG ->> UP: barrier(epoch+1)
+    end
+    deactivate CP
+```
+
+### State Storage
+
+The state of UDAF is managed by the compute node as a single encoded BYTEA value.
+
+Currently, each aggregate operator has a **result table** to store the aggregate result. For most of our built-in aggregate functions, they have the same output as their state, so the result table is actually being used as the state table. However, for general UDAFs, their state may not be the same as their output. Such functions are not supported for now.
+
+Therefore, we propose to **transform the result table into state table**. The content of the table remains the same for existing functions. But for new functions whose state is different from output, only the state is stored. The output can be computed from the state when needed.
+
+For example, given the input:
+
+| Epoch | op   | id (pk) | v0   | w0   | v1    |
+| ----- | ---- | ------- | ---- | ---- | ----- |
+| 1     | +    | 0       | 1    | 2    | false |
+| 2     | U-   | 0       | 1    | 2    | false |
+| 2     | U+   | 0       | 2    | 1    | true  |
+
+The new **state table** (derived from old result table) of the agg operator would be like:
+
+| Epoch | id   | sum(v0) | bool_and(v1)        | weighted_avg(v0, w0) | max(v0)*   |
+| ----- | ---- | ------- | ------------------- | -------------------- | ---------- |
+| 1     | 0    | sum = 1 | false = 1, true = 0 | encode(1,2) = b'XXX' | output = 1 |
+| 2     | 0    | sum = 2 | false = 0, true = 1 | encode(2,1) = b'YYY' | output = 2 |
+
+* For **append-only** aggregate functions (e.g. max, min, first, last, string_agg...), their states are all input values maintained in seperate "materialized input" tables. For backward compatibility, their values in the state table are still aggregate results.
```

Comment:
non-append-only?

Reply:
Non-append-only (retractable) functions are described above as a general case. This section is just a note for the special case.